gapis/api/vulkan/graph_visualization.go

```diff
@@ -66,60 +75,127 @@ func getSubCommandHierarchyNames() *api.HierarchyNames {
+func (builder *labelForVulkanCommands) GetCommandLabel(command api.Cmd, cmdId uint64) *api.Label {
+	label := &api.Label{}
 	commandName := command.CmdName()
-	label := ""
-	if commandBuffer := getCommandBuffer(command); commandBuffer != "" {
-		if _, ok := builder.labelToHierarchy[commandBuffer]; !ok {
-			builder.labelToHierarchy[commandBuffer] = &api.Hierarchy{}
+	if commandBufferId, ok := getCommandBufferId(command); ok {
+		if _, ok := builder.commandBufferIdToHierarchy[commandBufferId]; !ok {
+			builder.commandBufferIdToHierarchy[commandBufferId] = &api.Hierarchy{}
+			builder.commandBufferIdToOrderNumber[commandBufferId] = len(builder.commandBufferIdToOrderNumber) + 1
 		}
-		hierarchy := builder.labelToHierarchy[commandBuffer]
-		label += commandBuffer + "/"
-		label += getLabelFromHierarchy(commandName, commandHierarchyNames, hierarchy)
-		label += fmt.Sprintf("%s_%d", commandName, commandNodeId)
+		hierarchy := builder.commandBufferIdToHierarchy[commandBufferId]
+		label.PushBack(COMMAND_BUFFER, builder.commandBufferIdToOrderNumber[commandBufferId])
+		label.PushBackLabel(getLabelFromHierarchy(commandName, commandHierarchyNames, hierarchy))
+		label.PushBack(commandName, int(cmdId))
 	} else {
-		label += fmt.Sprintf("%s_%d", commandName, commandNodeId)
+		label.PushBack(commandName, int(cmdId))
 	}
+	checkDebugMarkers(builder, label)
 	return label
 }
 
-func (builder *labelForVulkanCommands) GetSubCommandLabel(index api.SubCmdIdx, commandName, subCommandName string) string {
-	label := commandName
-	subCommandIndexName := commandName
+func (builder *labelForVulkanCommands) GetSubCommandLabel(index api.SubCmdIdx, commandName string, cmdId uint64, subCommandName string) *api.Label {
+	label := &api.Label{}
+	label.PushBack(commandName, int(cmdId))
+	var subCommandIndexName bytes.Buffer
+	fmt.Fprintf(&subCommandIndexName, "%s_%d", commandName, cmdId)
 	for i := 1; i < len(index); i++ {
-		subCommandIndexName += fmt.Sprintf("/%d", index[i])
+		fmt.Fprintf(&subCommandIndexName, "/%d", index[i])
 		if i+1 < len(index) {
-			if hierarchyLabel, ok := builder.subCommandIndexNameToHierarchyLabel[subCommandIndexName]; ok {
-				label += "/" + hierarchyLabel
+			if hierarchyLabel, ok := builder.subCommandIndexNameToHierarchyLabel[subCommandIndexName.String()]; ok {
+				label.PushBackLabel(hierarchyLabel)
 			} else {
-				label += fmt.Sprintf("/%d", index[i])
+				label.PushBack("", int(index[i]))
 			}
 		}
 	}
-	if _, ok := builder.labelToHierarchy[label]; !ok {
-		builder.labelToHierarchy[label] = &api.Hierarchy{}
+	temporalLabelAsAString := label.GetLabelAsAString()
+	if _, ok := builder.labelAsAStringToHierarchy[temporalLabelAsAString]; !ok {
+		builder.labelAsAStringToHierarchy[temporalLabelAsAString] = &api.Hierarchy{}
 	}
-	hierarchy := builder.labelToHierarchy[label]
+	hierarchy := builder.labelAsAStringToHierarchy[temporalLabelAsAString]
```

Comment:
In general this pattern of inserting a default value into the map when the key is missing can be re-expressed as
```
hierarchy, ok := builder.labelAsAStringToHierarchy[temporalLabelAsAString]
if !ok {
	hierarchy = &api.Hierarchy{}
	builder.labelAsAStringToHierarchy[temporalLabelAsAString] = hierarchy
}		
```
This removes one of the map lookups.

Reply:
Thank you. Yes , using this I am removing one look up in the map. I changed it for the two cases.